src/ctap/credential_id.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::crypto_wrapper::{aes256_cbc_decrypt, aes256_cbc_encrypt, PrivateKey};
+use super::data_formats::CredentialProtectionPolicy;
+use super::status_code::Ctap2StatusCode;
+use super::{cbor_read, cbor_write};
+use crate::api::key_store::KeyStore;
+#[cfg(feature = "ed25519")]
+use crate::ctap::data_formats::EDDSA_ALGORITHM;
+use crate::ctap::data_formats::{extract_byte_string, extract_map, ES256_ALGORITHM};
+use crate::env::Env;
+use alloc::vec::Vec;
+use core::convert::{TryFrom, TryInto};
+use crypto::hmac::{hmac_256, verify_hmac_256};
+use crypto::sha256::Sha256;
+use sk_cbor::{cbor_map_options, destructure_cbor_map};
+
+pub const LEGACY_CREDENTIAL_ID_SIZE: usize = 112;
+// CBOR credential IDs consist of
+// - 1   byte : version number
+// - 16  bytes: initialization vector for AES-256,
+// - 240 bytes: encrypted block of the key handle cbor,
+// - 32  bytes: HMAC-SHA256 over everything else.
+#[cfg(test)]
+pub const CBOR_CREDENTIAL_ID_SIZE: usize = 241;
+pub const MAX_CREDENTIAL_ID_SIZE: usize = 241;
+pub const MIN_CREDENTIAL_ID_SIZE: usize = 112;
+
+pub const ECDSA_CREDENTIAL_ID_VERSION: u8 = 0x01;
+#[cfg(feature = "ed25519")]
+pub const ED25519_CREDENTIAL_ID_VERSION: u8 = 0x02;
+pub const CBOR_CREDENTIAL_ID_VERSION: u8 = 0x03;
+
+pub const MAX_PADDING_LENGTH: u8 = 0xBF;
+
+pub(crate) struct CredentialId {
+    pub private_key: PrivateKey,
+    pub rp_id_hash: [u8; 32],
+    pub cred_protect_policy: Option<CredentialProtectionPolicy>,
+}
+
+fn decrypt_legacy_key_handle(
+    env: &mut impl Env,
+    bytes: &[u8],
+    algorithm: i64,
+) -> Result<Option<CredentialId>, Ctap2StatusCode> {
+    let aes_enc_key = crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+    let plaintext = aes256_cbc_decrypt(&aes_enc_key, bytes, true)?;
+    if plaintext.len() != 64 {
+        return Ok(None);
+    }
+    let private_key = if let Some(key) = match algorithm {
+        ES256_ALGORITHM => PrivateKey::new_ecdsa_from_bytes(&plaintext[..32]),
+        #[cfg(feature = "ed25519")]
+        EDDSA_ALGORITHM => PrivateKey::new_ed25519_from_bytes(&plaintext[..32]),
+        _ => None,
+    } {
+        key
+    } else {
+        return Ok(None);
+    };
+    Ok(Some(CredentialId {
+        private_key,
+        rp_id_hash: plaintext[32..64].try_into().unwrap(),
```

Comment:
We usually went with `*array_ref![plaintext, 32, 32]` for these conversions, but we can break our usual style if there is good reason to?

Reply:
I think `arrayref` was used because back then it wasn't possible to convert slices to arrays. I think it's better to use what the language provides than rely on a third-party crate.